### PR TITLE
Avoid use of gRPC FeeData type in resource usage estimate

### LINKED ISSUE #1517: Remove unneeded use of gRPC types in fee calculation, starting with CryptoTransfer
**Summary**
Both the fee calculation utilities in _hapi-utils/_ and the resource usage estimators in _hapi-fees/_ use the gRPC `FeeData` structure extensively. 

**Possible resolution**
It will be simpler and much, much faster to use a mutable POJO to accumulate resource usage estimates.

### PR DESCRIPTION
**Pull request by tinker-michaelj**
Sonar passed with no material new code smells [here](https://sonarcloud.io/dashboard?branch=01517-M0150-AvoidFeeDataGrpcTypes&id=com.hedera.hashgraph%3Ahedera-services).

**Related issue(s)**:
- Closes #1517

**Summary of the change**:
- Create a mutable `UsageAccumulator` pojo as an alternative to the gRPC `FeeData` type currently used to express a resource usage estimate.
- Refactor `CryptoTransfer` and `ConsensusSubmitMessage` to use this new type via methods in the `CryptoOpsUsage` and `ConsensusOpsUsage` helper classes.
- For now, extend the `TxnAccessor` with the additional metadata needed to estimate `CryptoTransfer` and `ConsensusSubmitMessage` usage efficiently.
    * In the future, refactor the `ExpandHandleSpan` to track a parent object containing the `TxnAccessor` and such metadata as separate fields.

**Applicable documentation**
- [x] Javadoc

### CODE REVIEW

File: hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		final long networkFeeTinycents = networkFeeInTinycents(usage, prices.getNetworkdata());
+		final long nodeFeeTinycents = nodeFeeInTinycents(usage, prices.getNodedata());
+		final long serviceFeeTinycents = serviceFeeInTinycents(usage, prices.getServicedata());
+
+		final long networkFee = tinycentsToTinybars(networkFeeTinycents, rate) * multiplier;
+		final long nodeFee = tinycentsToTinybars(nodeFeeTinycents, rate) * multiplier;
+		final long serviceFee = tinycentsToTinybars(serviceFeeTinycents, rate) * multiplier;
+
+		if (networkFee < 0 || nodeFee < 0 || serviceFee < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
```
Comment (ljianghedera) [inline hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java:46]: Does `0` (zero) belong to positive numbers?

File: hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+		}
+		base += a;
+		if (base < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
+		}
+		base += b;
+		if (base < 0) {
+			throw new IllegalArgumentException(POSITIVITY_ERROR);
+		}
```
Comment (ljianghedera) [inline hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java:120]: nit: though unlikely, but If any of these cases happen, there must be some overflow. Will it be better to use a different message to indicate this case?

**Review by Neeharika-Sompalli — APPROVED** [review]
Overall LGTM

File: hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java (hunk 1 above)
Comment (qnswirlds) [reply hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java:46]: Good point @ljianghedera, we should rename it to `NON_NEGATIVITY_ERROR` and change the string content.

File: hedera-node/src/main/java/com/hedera/services/utils/SignedTxnAccessor.java [hunk 3]
```diff
@@ -98,6 +111,14 @@ public SignedTxnAccessor(byte[] signedTxnWrapperBytes) throws InvalidProtocolBuf
 		numSigPairs = sigMap.getSigPairCount();
 		utf8MemoBytes = StringUtils.getBytesUtf8(memo);
 		memoHasZeroByte = Arrays.contains(utf8MemoBytes, (byte) 0);
+
+		getFunction();
```
Comment (ljianghedera) [inline hedera-node/src/main/java/com/hedera/services/utils/SignedTxnAccessor.java:115]: Does this line serve any purpose?

**Review by qnswirlds — APPROVED** [review]
Overall LGTM. But have we run CI _without_ the changes in `test-clients`? If not, I would suggest we do so with the hope that CI has enough coverage for those fees and that we didn't change any logics (by not breaking CI) in this refactoring.

**Review by ljianghedera — APPROVED** [review]
LGTM.

File: hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java (hunk 2 above)
Comment (tinker-michaelj) [reply hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java]: True! Renamed to `OVERFLOW_ERROR` and changed message to,
```
A fee calculation step overflowed; the operation cannot be priced, and therefore cannot be performed
```

File: hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+import com.hedera.services.usage.state.UsageAccumulator;
+import com.hederahashgraph.api.proto.java.ExchangeRate;
+import com.hederahashgraph.api.proto.java.FeeComponents;
+import com.hederahashgraph.api.proto.java.FeeData;
+import com.hederahashgraph.fee.FeeBuilder;
+import com.hederahashgraph.fee.FeeObject;
+
+import static com.hedera.services.usage.SingletonEstimatorUtils.ESTIMATOR_UTILS;
+import static com.hederahashgraph.fee.FeeBuilder.FEE_DIVISOR_FACTOR;
+
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply hapi-fees/src/main/java/com/hedera/services/calc/OverflowCheckingCalc.java:32]: Added!

File: hedera-node/src/main/java/com/hedera/services/utils/SignedTxnAccessor.java (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/utils/SignedTxnAccessor.java:115]: It sets the `function` field so we can decide next whether to compute the `CryptoTransfer` or `ConsensusSubmitMessage` usage metadata.

**Comment by tinker-michaelj** [comment]
> Overall LGTM. But have we run CI _without_ the changes in `test-clients`? If not, I would suggest we do so with the hope that CI has enough coverage for those fees and that we didn't change any logics (by not breaking CI) in this refactoring.

Good point...before switching the EET `HapiCryptoTransfer.feeFor()` and `HapiSubmitMessage.feeFor()` over to the new usage estimators, I ran e.g. the `SubmitMessageSuite` and `TokenTransactSpecs` locally using the old estimators.

When migrating a more complicated estimator (e.g. `CryptoUpdate`), this will be even more important. 👍

File: hedera-node/src/main/java/com/hedera/services/utils/SignedTxnAccessor.java (hunk 3 above)
Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/utils/SignedTxnAccessor.java:115]: Yeah. I was thinking it should be `function = getFunction(); ` , but `getFunction();` here already serves the purpose.

**Pull request merged**